### ISSUE TITLE
FTP.ftp_retr should check openForReading for IsADirectoryError

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5441]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#5441|
|Type|defect|
|Created|2011-12-26 17:09:12Z|
|Branch|https://github.com/twisted/twisted/tree/ftpretr-directory-5441|

Right now `FTP.ftp_retr` is checking `openForReading` call for `IsNotADirectory`, while it should check for `IsADirectory` as `shell.openForReading` is raising `IsADirectory` exception.

I will attach a branch with the fix.

Attachments:

* [5441.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a17/a170c8689e2340bc3ef4b1210380b2559aabd20a/2d3c69bc203ff7404471a47a06ec980d64823c86.diff) (2025 bytes) - added by adiroiban on 2011-12-27 09:43:13Z - 
* [5441.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a17/a170c8689e2340bc3ef4b1210380b2559aabd20a/2d099e35f5428d3c3755a9f797221f6c695b3a41.diff) (1702 bytes) - added by adiroiban on 2012-01-15 05:49:01Z - 
* [5441.3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a17/a170c8689e2340bc3ef4b1210380b2559aabd20a/167080c1b95363ee63968f8cc747148633da329f.diff) (2759 bytes) - added by adiroiban on 2012-01-15 14:00:01Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5441 5441
type__defect defect
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_ftpretr_directory_5441 branches/ftpretr-directory-5441
branch_author__jesstess jesstess
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1324919352000000 1324919352000000
changetime__1326652275000000 1326652275000000
version__None None
owner__jesstess jesstess
cc__itamarst cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban set owner to @adiroiban|
|-|-|

code is here: https://code.launchpad.net/~adiroiban/twisted/5441
MP is here: https://code.launchpad.net/~adiroiban/twisted/5441/+merge/86916

Please let me know if you want to merge/land request in another format.

Cheers

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Thanks for your contribution. Our development process prefers that you supply a patch against trunk that we can review, see TwistedDevelopment#SubmittingaPatch for more information. You can also read the [Coding standard](http://twistedmatrix.com/trac/browser/trunk/doc/core/development/policy/coding-standard.xhtml?format=raw) and other [policies](http://twistedmatrix.com/trac/wiki/TwistedDevelopment#Policies) for more information.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

thijs,

Please don't reject legitimate contributions unless you've carefully inspected them.

A launchpad merge proposal is trivial to convert to a plain text diff, so we can accept them.  (Some reviewers might actually prefer them.)

If you're not familiar with bzr and launchpad, there's a button, "diff against target".  which will take you here: https://code.launchpad.net/~adiroiban/twisted/5441/+merge/86916#review-diff

and you can click "download diff" to get a plain-text diff that should apply to trunk.

Sorry for the delay, adiroiban!  I'm putting this back into review, so hopefully someone will get to it soon...

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Hi,

I have attached a diff generated using svn.

The Launchpad Merge Proposal page is done against twisted/trunk and contains a diff web view and a link for downloading a patch file. My previous changes were accepted based on bzr branches.

Cheers,
Adi

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

Replying to [glyph](#note_4):
> thijs,
> 
> Please don't reject legitimate contributions unless you've carefully inspected them.
> 
> A launchpad merge proposal is trivial to convert to a plain text diff, so we can accept them.  (Some reviewers might actually prefer them.)
> 
> If you're not familiar with bzr and launchpad, there's a button, "diff against target".  which will take you here: https://code.launchpad.net/~adiroiban/twisted/5441/+merge/86916#review-diff
> 
> and you can click "download diff" to get a plain-text diff that should apply to trunk.

Alright, I wasn't aware of that. I guess the TwistedDevelopment page should point this out then, because I couldn't find anything official about the fact it's ok to use github/bitbucket/bzr/etc and let the core developers figure out how to generate a plain text diff (no matter how easy it is). The development page also suggest: "If you don't have commit access to Twisted, please submit changesets in diff -u format to the ticket tracker. "

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [thijs](#note_6):

> Alright, I wasn't aware of that. I guess the TwistedDevelopment page should point this out then, because I couldn't find anything official about the fact it's ok to use github/bitbucket/bzr/etc and let the core developers figure out how to generate a plain text diff (no matter how easy it is). The development page also suggest: "If you don't have commit access to Twisted, please submit changesets in diff -u format to the ticket tracker. "

This is true, and more or less correct; ideally, contributors would also provide a patch.  It's just that saying that they should generate a patch doesn't really constitute a full review :).  By all means, feel free to stick a _comment_ on a ticket saying that a patch may accelerate the process of getting it reviewed; it should just remain in the review state in case another twisted core dev is comfortable with the contributor's selected method of contribution.

-glyph

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_10"></a>|@jesstess set owner to @adiroiban|
|-|-|

Thanks for the bug report and patch, adiroiban! This patch looks good; I just have a couple of cosmetic/documentation comments:

 * Tests should start with "test_", followed by a camel-case name where the first word starts lower-cased, ```test_downloadFolder``` in this case. (```test_ftp.py``` is quite old and has a bunch of code that hasn't been updated to meet the current testing standard)
 * ```test_downloadFolder``` needs a docstring.
 * "Downloading a file should not succeed" ==> "Downloading a folder should not succeed"
 * While you're touching ```ftp_RETR```, can you give it a docstring?

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

Thanks for the review!

I have attached a new patch and also updated the bzr branch.
The patch is built on top of the previous patch. Please let me know if a full patch against the base or latest version is required.

Cheers,
Adi

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

All things equal, I think most people would find a patch against trunk more useful than a patch against a bzr branch.  Using the latter involves applying multiple patches or getting the bzr branch, and once you have the bzr branch you don't need the patch file anymore.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

Hi,

I have attached a new diff against the latest trunk.

Cheers,
Adi

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_14"></a>|@jesstess commented|
|-|-|

(In [33433]) Branching to 'ftpretr-directory-5441'

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_16"></a>|@jesstess commented|
|-|-|

(In [33434]) Apply patch 5441.3.diff by adiroiban.

With some tiny cosmetic tweaks (docstring """ on lines by themselves).

refs [#5441](https://github.com/twisted/twisted/issues/5441)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_17"></a>|@jesstess commented|
|-|-|

Thanks for the updated patch, adiroiban. [Buildbot results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ftpretr-directory-5441) look good, so I'll go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_18"></a>|@jesstess set status to `closed`|
|-|-|

(In [33435]) Merge ftpretr-directory-5441

Author: adiroiban
Reviewer: jesstess
Fixes: [#5441](https://github.com/twisted/twisted/issues/5441)

Correct a check in FTP.ftp_retr from IsNotADirectoryError to
IsADirectoryError. Now calling RETR on a directory won't be logged as
an unexpected error.
